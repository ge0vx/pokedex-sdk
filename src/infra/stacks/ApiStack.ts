import { Stack, StackProps } from "aws-cdk-lib";
import { Cors, LambdaIntegration, ResourceOptions, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";

interface ApiStackProps extends StackProps {
    favPokemonIntegration: LambdaIntegration
}

export class ApiStack extends Stack {

    constructor(scope: Construct, id: string, props?: ApiStackProps){
        super(scope, id, props);

        const api = new RestApi(this, 'favPokemonApi');

        const optionsWithCors: ResourceOptions = {
            defaultCorsPreflightOptions: {
                allowOrigins: Cors.ALL_ORIGINS,
                allowMethods: Cors.ALL_METHODS,
            }
        }

        const favPokemonResource = api.root.addResource('favPokemon', optionsWithCors);
        favPokemonResource.addMethod('GET', props.favPokemonIntegration)
        favPokemonResource.addMethod('POST', props.favPokemonIntegration)
        
    }
}