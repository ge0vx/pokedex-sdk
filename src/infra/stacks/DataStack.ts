import { Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";

export class DataStack extends Stack {

    public readonly favPokemonTable: ITable;

    constructor(scope: Construct, id: string, props?: StackProps){
        super(scope, id, props)

        this.favPokemonTable = new Table(this, 'favPokemonTable', {
            partitionKey : {
                name: 'id',
                type: AttributeType.STRING
            },
            tableName : 'favPokemonTable'
        })
    }
}