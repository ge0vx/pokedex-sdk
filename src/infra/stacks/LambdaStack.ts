import { Stack, StackProps } from "aws-cdk-lib";
import { LambdaIntegration } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import { join } from "path";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";

interface LambdaStackProps extends StackProps {
    favPokemonTable: ITable
}

export class LambdaStack extends Stack {

    public readonly favPokemonIntegration: LambdaIntegration;
    public readonly pokemonLamdaIntegration: LambdaIntegration;

    constructor(scope: Construct, id: string, props?: LambdaStackProps){
        super(scope, id, props);

        const runtime = Runtime.NODEJS_18_X;
        const handlerFileName = 'handler';

        const favPokemonLambda = new NodejsFunction(this, 'favPokemonLambda',{
            runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'favPokemon', 'handler.ts'),
            environment: {
                TABLE_NAME: props.favPokemonTable.tableName
            }
        })

        favPokemonLambda.addToRolePolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [props.favPokemonTable.tableArn],
            actions:[
                'dynamodb:PutItem',
                'dynamodb:Scan'
            ]
        }))

        const pokemonLambda = new NodejsFunction(this, 'PokemonLamda', {
            runtime: runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'pokemon', 'handler.ts'),
        })
        
        this.favPokemonIntegration = new LambdaIntegration(favPokemonLambda)
        this.pokemonLamdaIntegration = new LambdaIntegration(pokemonLambda)
    }
}