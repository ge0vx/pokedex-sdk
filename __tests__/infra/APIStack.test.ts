import { App } from "aws-cdk-lib";
import { LambdaStack } from "../../src/infra/stacks/LambdaStack";
import { Template } from "aws-cdk-lib/assertions";
import { DataStack } from "../../src/infra/stacks/DataStack";
import { ApiStack } from "../../src/infra/stacks/ApiStack";

describe('Initial test suite', ()=>{

    let apiStackTemplate: Template;

    beforeAll(()=>{
        const testApp = new App({
            outdir: 'cdk.out'
        })

        const dataStack = new DataStack(testApp, 'DataStackFavPokemon');
        const lambdaStack = new LambdaStack(testApp, 'LambdaStackFavPokemon', {
            favPokemonTable : dataStack.favPokemonTable
        });
        const apiStack = new ApiStack(testApp, 'ApiStackFavPokemon', {
            favPokemonIntegration: lambdaStack.favPokemonIntegration,
            pokemonLambdaIntegration: lambdaStack.pokemonLamdaIntegration
        })

        apiStackTemplate = Template.fromStack(apiStack);
    })

    test('ApiGateway properties', ()=>{
        apiStackTemplate.hasResourceProperties('AWS::ApiGateway::RestApi', {
            Name: "favPokemonApi"
        })
    });

    test('ApiGateway Rest API service snapshot', ()=>{
        expect(apiStackTemplate.toJSON()).toMatchSnapshot();
    })
    
});