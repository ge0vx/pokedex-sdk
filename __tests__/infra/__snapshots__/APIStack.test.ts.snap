// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suite ApiGateway Rest API service snapshot 1`] = `
{
  "Outputs": {
    "favPokemonApiEndpoint89992AA8": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "favPokemonApi8D02935E",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "favPokemonApiDeploymentStageprodA65E66FF",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "favPokemonApi8D02935E": {
      "Properties": {
        "Name": "favPokemonApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "favPokemonApiAccount68C42CB1": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "favPokemonApi8D02935E",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "favPokemonApiCloudWatchRole9D5C2DEB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "favPokemonApiCloudWatchRole9D5C2DEB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "favPokemonApiDeployment74C1D062d61af481c72453c8fc699b784558f6e0": {
      "DependsOn": [
        "favPokemonApifavPokemonGETE8E45793",
        "favPokemonApifavPokemonOPTIONS55B4EDC5",
        "favPokemonApifavPokemonPOSTAA9F0B8C",
        "favPokemonApifavPokemonF31CDD74",
        "favPokemonApipokemonGET7ADF943C",
        "favPokemonApipokemonOPTIONS7C9E7188",
        "favPokemonApipokemonD494A5F6",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "favPokemonApiDeploymentStageprodA65E66FF": {
      "DependsOn": [
        "favPokemonApiAccount68C42CB1",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "favPokemonApiDeployment74C1D062d61af481c72453c8fc699b784558f6e0",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "favPokemonApifavPokemonF31CDD74": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "favPokemonApi8D02935E",
            "RootResourceId",
          ],
        },
        "PathPart": "favPokemon",
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "favPokemonApifavPokemonGETApiPermissionApiStackFavPokemonfavPokemonApi46EBFF55GETfavPokemon3563314A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/",
              {
                "Ref": "favPokemonApiDeploymentStageprodA65E66FF",
              },
              "/GET/favPokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApifavPokemonGETApiPermissionTestApiStackFavPokemonfavPokemonApi46EBFF55GETfavPokemon2180194C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/test-invoke-stage/GET/favPokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApifavPokemonGETE8E45793": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "favPokemonApifavPokemonF31CDD74",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "favPokemonApifavPokemonOPTIONS55B4EDC5": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "favPokemonApifavPokemonF31CDD74",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "favPokemonApifavPokemonPOSTAA9F0B8C": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "favPokemonApifavPokemonF31CDD74",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "favPokemonApifavPokemonPOSTApiPermissionApiStackFavPokemonfavPokemonApi46EBFF55POSTfavPokemonFDBE8EEF": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/",
              {
                "Ref": "favPokemonApiDeploymentStageprodA65E66FF",
              },
              "/POST/favPokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApifavPokemonPOSTApiPermissionTestApiStackFavPokemonfavPokemonApi46EBFF55POSTfavPokemon7AB223EE": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttfavPokemonLambdaBFB7282AArnCA5AD750",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/test-invoke-stage/POST/favPokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApipokemonD494A5F6": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "favPokemonApi8D02935E",
            "RootResourceId",
          ],
        },
        "PathPart": "pokemon",
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "favPokemonApipokemonGET7ADF943C": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttPokemonLamda31F64495ArnE7A94F90",
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "favPokemonApipokemonD494A5F6",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "favPokemonApipokemonGETApiPermissionApiStackFavPokemonfavPokemonApi46EBFF55GETpokemon56034B92": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttPokemonLamda31F64495ArnE7A94F90",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/",
              {
                "Ref": "favPokemonApiDeploymentStageprodA65E66FF",
              },
              "/GET/pokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApipokemonGETApiPermissionTestApiStackFavPokemonfavPokemonApi46EBFF55GETpokemon4A0D77EE": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::ImportValue": "LambdaStackFavPokemon:ExportsOutputFnGetAttPokemonLamda31F64495ArnE7A94F90",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "favPokemonApi8D02935E",
              },
              "/test-invoke-stage/GET/pokemon",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "favPokemonApipokemonOPTIONS7C9E7188": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "favPokemonApipokemonD494A5F6",
        },
        "RestApiId": {
          "Ref": "favPokemonApi8D02935E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
